Made changes to compile and run on Linux with gcc.

sts1skj opened the pull request:
1.  Created CMakeLists.txt to build with CMake on Linux.  It might also
work on other systems.

2.  stricmp and strnicmp were being used; they are missing from
gcc's standard library implementation.  (This is a well-known and
long-standing bug in gcc.)  So I added src/utils/StringStuff.{cpp,h}
with implementations of case-insensitive string compare functions
and changed all instances of stricmp and strnicmp to use them.

3.  Added StringStuff to the VC project files so hopefully it'll still
build on that tool (which I do not have).

4.  Fixed a variety of minor bugs where the code was not standard
C++, but presumably VC was permissive enough that it compiled anyway.
Hopefully VC will accept these changes; I don't have VC.  The most
common bug was assigning a wxString to a std::string in an old-style
initializer; switching to the new initializer style fixed it.

5.  Added a few missing '#include' statements.

6.  Made some no-effect changes to silence compiler warnings, such as
adding "default: break;" to switch statements where some of the values
were not handled.

7.  gcc is missing acosf, cosf, sinf, atan2f, and fabsf.  (This is
a well-known and long-standing bug in gcc.)  Luckily, these are
just alternative names for existing functions, so I switched to the
other names.

8.  There were many locations where BS&OS tried to open files
with filenames where the path separator character was backslash.
I changed them to use forward slash instead.  This shouldn't affect
Windows execution, because Windows accepts either type of slash as a
path separator.  I tried my best to not let the slash style propagate
into data files, but I may not have been successful.

9.  There were a few locations where filenames were being converted to
lowercase before the file was opened.  As Linux is case sensitive, this
didn't work.  So I changed all of these instances so that the filename
is not converted.

10.  On Linux, the wxWidgets headers "#include <X11/X.h>", which is an
ancient legacy header file containing a large number of poorly-named
"#define" statements.  I feel that this is a bug in wxWidgets.  There
was a name clash with a function name in the FBX SDK.  So I moved the
"#include <fbxsdk.h>" line from FBXWrangler.h into FBXWrangler.cpp
so that FBX and wxWidgets headers would never be included by the
same file.  To keep FBXWrangler working, I needed to move more of
the implementation details from the header into the source file.
This necessitated employing the private-implementation idiom.

11.  The C integer-absolute-value function "abs" was being used in
TriFile.cpp on floating point values.  gcc produced a warning, and
I judged this highly likely to be a bug, so I fixed it by replacing
"abs" with "std::abs".

12.  Linux doesn't have a registry, so I used "#ifdef _WINDOWS"
around the handful of blocks of registry code.

13.  In BodySlideApp.h, there was some peculiar code for doing something
with number-key presses and Windows events.  I used "#ifdef _WINDOWS"
to disable this code for linux.

14.  A few files were missing trailing newline characters.
Linux doesn't like that, so I added them.

15.  Switched from old-style cast to "reinterpret_cast" a few places
in NormalsGenDialog to silence compiler complaints.

16.  "_snprintf_s" was used in a few places.  It's a Windows-only
function.  I replaced it with snprintf, which does the same thing
only better.

17.  The extension-loading code in GLExtensions.{h,cpp} used
wglGetProcAddress, which is Windows-specific.  I couldn't figure out
how to convert the code to Linux, so, for Linux, GL extensions are
now loaded by GLEW.  This new code would probably work fine on other
systems too, but it would introduce a dependence on the GLEW library.

18.  Three OpenGL API calls had capitalization errors.  I fixed them.

After all this, BodySlide and Outfit Studio build and run, apparently
correctly, on my Linux system.

sts1skj committed:
Made changes to compile and run on Linux with gcc.

1.  Created CMakeLists.txt to build with CMake on Linux.  It might also
work on other systems.

2.  stricmp and strnicmp were being used; they are missing from
gcc's standard library implementation.  (This is a well-known and
long-standing bug in gcc.)  So I added src/utils/StringStuff.{cpp,h}
with implementations of case-insensitive string compare functions
and changed all instances of stricmp and strnicmp to use them.

3.  Added StringStuff to the VC project files so hopefully it'll still
build on that tool (which I do not have).

4.  Fixed a variety of minor bugs where the code was not standard
C++, but presumably VC was permissive enough that it compiled anyway.
Hopefully VC will accept these changes; I don't have VC.  The most
common bug was assigning a wxString to a std::string in an old-style
initializer; switching to the new initializer style fixed it.

5.  Added a few missing '#include' statements.

6.  Made some no-effect changes to silence compiler warnings, such as
adding "default: break;" to switch statements where some of the values
were not handled.

7.  gcc is missing acosf, cosf, sinf, atan2f, and fabsf.  (This is
a well-known and long-standing bug in gcc.)  Luckily, these are
just alternative names for existing functions, so I switched to the
other names.

8.  There were many locations where BS&OS tried to open files
with filenames where the path separator character was backslash.
I changed them to use forward slash instead.  This shouldn't affect
Windows execution, because Windows accepts either type of slash as a
path separator.  I tried my best to not let the slash style propagate
into data files, but I may not have been successful.

9.  There were a few locations where filenames were being converted to
lowercase before the file was opened.  As Linux is case sensitive, this
didn't work.  So I changed all of these instances so that the filename
is not converted.

10.  On Linux, the wxWidgets headers "#include <X11/X.h>", which is an
ancient legacy header file containing a large number of poorly-named
"#define" statements.  I feel that this is a bug in wxWidgets.  There
was a name clash with a function name in the FBX SDK.  So I moved the
"#include <fbxsdk.h>" line from FBXWrangler.h into FBXWrangler.cpp
so that FBX and wxWidgets headers would never be included by the
same file.  To keep FBXWrangler working, I needed to move more of
the implementation details from the header into the source file.
This necessitated employing the private-implementation idiom.

11.  The C integer-absolute-value function "abs" was being used in
TriFile.cpp on floating point values.  gcc produced a warning, and
I judged this highly likely to be a bug, so I fixed it by replacing
"abs" with "std::abs".

12.  Linux doesn't have a registry, so I used "#ifdef _WINDOWS"
around the handful of blocks of registry code.

13.  In BodySlideApp.h, there was some peculiar code for doing something
with number-key presses and Windows events.  I used "#ifdef _WINDOWS"
to disable this code for linux.

14.  A few files were missing trailing newline characters.
Linux doesn't like that, so I added them.

15.  Switched from old-style cast to "reinterpret_cast" a few places
in NormalsGenDialog to silence compiler complaints.

16.  "_snprintf_s" was used in a few places.  It's a Windows-only
function.  I replaced it with snprintf, which does the same thing
only better.

17.  The extension-loading code in GLExtensions.{h,cpp} used
wglGetProcAddress, which is Windows-specific.  I couldn't figure out
how to convert the code to Linux, so, for Linux, GL extensions are
now loaded by GLEW.  This new code would probably work fine on other
systems too, but it would introduce a dependence on the GLEW library.

18.  Three OpenGL API calls had capitalization errors.  I fixed them.

After all this, BodySlide and Outfit Studio build and run, apparently
correctly, on my Linux system.

ousnius commented:
@sts1skj: Thanks for the detailed description of your changes. I've cloned your branch and the MSVC builds are successful. Basic tests are working but I need to test more.

I don't know if the void pointers thing in NormalsGenDialog is even working as intended (esp. on 64-bit builds), both before and after the changes, but it doesn't matter right now as it's not used.

There are one or two warnings I would like to fix. Additionally, I would like two of the forward slash changes reverted (in the .osp project files).

1. Output path is filled with forward slashes now. I don't want it to suddenly change with the next version, so the .osp should still store it as backslashes:
`<OutputPath>meshes/armor/New Outfit</OutputPath>`
2. When adding new slider data, the data link to the .osd is in forward slashes as well (after TEST.osd):
`<Data name="BaseShapeHRtest" target="BaseShapeHR" local="true">TEST.osd/BaseShapeHRtest</Data>`

Another thing in BodySlideApp.cpp is where it adds the .tri path to the NIF files (NiStringExtraData blocks). These strings need to be stored using backslashes in the NIF file (triPath and triPathTrimmed with regex).

Can you please give me permission to push to the PR branch so I can fix the warnings and anything else I might find?

sts1skj commented:
When you say "give me permission to push to the PR branch", are you referring to the "Allow edits from maintainers" checkbox?  That is already checked.  Please let me know if there's something else I need to do; I've never used github before.

ousnius committed:
Fix two warnings

- "size_t" instead of "int" for string length
- "int ret" in SavePreset was already defined in a Windows environment

ousnius commented:
@sts1skj Sorry, I had cloned the wrong branch. It's working.
I've edited my post above with some issues. Do you want to look at them?

sts1skj commented:
I will try to fix the two forward-slash problems you mentioned.

ousnius commented:
For BodySlide you will need to look at the `SliderSet::GetOutputFilePath` function and where it's used for the BodySlide build (wxRemoveFile uses it, and it's used for the tri file paths as well).

POMXARK commented:
Hi ousnius and Jeir. why not add a collision check? to check clipping, the mass of libraries, or is it difficult to implement (bullet3,FCL)??? to handle convergence and automatically adjust armor to the body? Is it worth trying to do this ???
In my large squat, fatal matches in the lower chest are possible, when it will be possible to automate the conversion process, whoever tried, I want to use macros for the program, as done in notepad ++

sts1skj committed:
Made all pathnames in files use backslashes.

1.  Added Readme-linux.txt, which describes the build process for Linux.

2.  Added ToOSSlashes and ToBackslashes functions to StringStuff.  They
convert slashes and backslashes to either the OS's path separator character
or backslashes, respectively.

3.  Added PathSepChar and PathSepStr constants to StringStuff.h.

4.  Made sure pathnames read in from files have backslashes converted
to the path separator character of the OS.  Made sure pathnames written
to files have forward slashes converted to backslashes.  This was done
for the following file types and data items:

material file: all texture pathnames
nif file:
	SetStringData (only for BODYTRI extra data on a shape)
	GetTextureSlot
	SetTextureSlot
slider set osp file:
	SliderSetInfo::SliderSet::DataFolder
	SliderSetInfo::SliderSet::OutputPath
	SliderSetInfo::SliderSet::Shape::DataFolder
	SliderSetInfo::SliderSet::Slider::Data
ref templates xml file: sourcefile attribute of RefTemplates::Template

I did not do this for the configuration file Config.xml, which has
a bunch, because I figured that file is never exchanged or used by
any other software.

5.  Generally made sure that both forward and back slashes are accepted
in pathnames internally, such as when splitting the filename off of
a path.

sts1skj committed:
Merge branch 'linux' of https://github.com/sts1skj/BodySlide-and-Outfit-Studio into linux

sts1skj committed:
Pulled in changes of c401910

sts1skj committed:
Merged changes of commit c401910

sts1skj commented:
I've checked in a set of changes that should fix the slash problem.  Now, pathnames are converted to backslashes when being written to nif, osp, material, and ref templates xml files.  Pathnames read from these files are converted to use the OS's path separator character.  So now internal pathname strings will generally use backslash on Windows and forward slash on all other OSes, except for string-literal pathnames that don't matter, which generally use forward slashes.

I tried to merge in the changes you made, Ousnius, to silence compiler warnings.  Hopefully I did that correctly; git is still somewhat mystifying to me.

I also added a Readme-linux.txt file with build instructions.

I didn't add code for making pathnames in Config.xml always use backslashes.  If you want me to make the changes for that, let me know.

I also examined the code for reading and writing the following file types to see if they store any pathnames, and I didn't find any:
shader file
tri file
tri head file
obj file
fbx file
slider set group xml
slider preset xml
slider categories xml
Please let me know if any of these types of files do contain pathnames or if I missed a file type.

sts1skj commented:
I need to do more testing.  Something isn't working right.

sts1skj committed:
Fixed a couple of linux-only bugs

1.  Two places where osp OutputPath were not being converted on input.
2.  "error" was not being set on failing to open a file.

sts1skj commented:
This should be ready now.  I've done a lot of testing, and everything I've tested works fine on Linux now.

ousnius committed:
Made texture cache case-insensitive

Textures didn't load/store right on Windows anymore.

ousnius committed:
Fixed _WINDOWS define

sts1skj commented:
Your commit 26c9e13 didn't break anything on the Linux side, as far as I can tell.

ousnius committed:
Fixed multiple slash issues

Forward slashes prevented slider data (.osd) from loading on Windows. Further changes done preemptively.

ousnius committed:
Restored continue (copy-paste error)

ousnius commented:
@sts1skj: If building a project in BodySlide, textures in the preview, loading -> saving -> loading a project in Outfit Studio works for you, all while keeping the slider data working, then I can merge the PR.

sts1skj committed:
Set slider background color in Outfit Studio for Linux.

wxWidgets for gtk2 and gtk3 has a well-known bug that makes it ignore some
"bg" properties in xrc files.  This results in some widgets having
system-defined background colors.  In Outfit Studio, there's just one
place where this is a problem: the background colors of sliders.  This
change sets the background color in the code.

sts1skj committed:
Merge branch 'linux' of https://github.com/sts1skj/BodySlide-and-Outfit-Studio into linux

sts1skj commented:
@ousnius: I added one more tiny change.  Hopefully it has no effect on Windows; it just sets the background color of sliders in Outfit Studio to what it would be anyway.

I've tested your latest commit ec64a13, and everything works on Linux according to the tests you specified.

sts1skj commented:
@ousnius: I added one more tiny change.  Hopefully it has no effect on Windows; it just sets the background color of sliders in Outfit Studio to what it would be anyway.

I've tested your latest commit ec64a13, and everything works on Linux according to the tests you specified.

ousnius committed:
Expand on af22b5e

ousnius commented:
I've added another place where the slider pane background color was set. This should be ready to merge now.

sts1skj commented:
Everything looks good to me.  Go ahead.

ousnius committed:
More use of PathSepChar

ousnius commented:
@sts1skj Thank you for the work on this!

sts1skj commented:
Thank you.
